File: 0189-rotate-array/0189-rotate-array.py
class Solution(object):
    def rotate(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: None Do not return anything, modify nums in-place instead.
        """

        dp = collections.deque(nums)

        for i in range(k):
            temp = dp.pop()
            dp.appendleft(temp)

        for i in range(len(nums)):
            nums[i] = dp[i]

        # k = k%len(nums)
        # nums[:] = nums[-k:] + nums[:-k]

File: 0189-rotate-array/0189-rotate-array.cpp
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        int n = nums.size();
        k %= n; // same as Python's k = k % len(nums)
        if (k == 0) return;

        // Reverse whole array
        std::reverse(nums.begin(), nums.end());

        // Reverse first k elements
        std::reverse(nums.begin(), nums.begin() + k);

        // Reverse the rest
        std::reverse(nums.begin() + k, nums.end());
    }
};